### DISCUSSION TITLE
automatic distribution of staking/delegation mana rewards

### DISCUSSION THREAD

**Discussion by Buddh1n1** [iotaledger/iota-core#705] [Ideas]
Instead of storing an endless history of validator performance and epoch mana reward stats, mana should be automatically assigned to the BIC vector after calculating rewards for an epoch. 
Delegators and validators would need to specify a block issuer during registration/delegation.

This also gets rid of the complexity for the process of claiming rewards.

To avoid mana becoming illiquid as a result, there is #703

**Comment by alexsporn** [comment]
Hi @Buddh1n1, thanks for the idea. This approach was in the initial discussions but was discarded due to the following isues:
- Updating the BIC for all users on all nodes at about the same time generates a high CPU and I/O load. We actually had this approach already on IOTA mainnet in the participation code, where we updated the votes on every milestone. With an increasing number of votes, the updating of the totals took longer and longer, until the nodes could not even keep up with the milestone intervals. Updating this cannot be done in parallel and requires a lock on the data structure, so the more users you have, the longer it will take.
- With the current approach we only mutate whatever really changed and keep track of the required performance indicators to be able to calculate the values when they are being used.
- We need to keep track of slot diffs (i.e. deltas on every slot) to be able to roll back the state in the case we need to fork.

This is the same reason why we don't update the generated mana on every output on every commitment and instead calculate the generated + decayed mana when an output is consumed.
